### DISCUSSION TITLE
Can't get rid of Original Error: *errors.errorString acme/autocert: missing server name

### DISCUSSION THREAD

**Discussion by ErrorProne** [gravitational/teleport#18802] [Q&A] [answered]
Hi there,

I'm trying to tinker around with teleport 11 on a test ubuntu instance in our DMZ (I pretty much followed this guide https://goteleport.com/docs/deploy-a-cluster/open-source/)
All traffic to the VM passes through the pfSense haproxy, for which I've tried nearly every variation to pass traffic on to the teleport VM. However I always run into this error:

```
Nov 27 00:42:57 teleport-bastion teleport[15578]: 2022-11-27T00:42:57Z DEBU [AUTH]      ClientCertPool -> cert(teleport.mytest.de issued by teleport.mytest.de:143385119287013782585507479881644659881) auth/middleware.go:674
Nov 27 00:42:57 teleport-bastion teleport[15578]: 2022-11-27T00:42:57Z WARN [ALPN:PROX] Failed to handle client connection. error:[
Nov 27 00:42:57 teleport-bastion teleport[15578]: ERROR REPORT:
Nov 27 00:42:57 teleport-bastion teleport[15578]: Original Error: *errors.errorString acme/autocert: missing server name
Nov 27 00:42:57 teleport-bastion teleport[15578]: Stack Trace:
Nov 27 00:42:57 teleport-bastion teleport[15578]:         github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:389 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
Nov 27 00:42:57 teleport-bastion teleport[15578]:         github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:327 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
Nov 27 00:42:57 teleport-bastion teleport[15578]:         runtime/asm_amd64.s:1594 runtime.goexit
Nov 27 00:42:57 teleport-bastion teleport[15578]: User Message: acme/autocert: missing server name] alpnproxy/proxy.go:335
```

I wonder how I can get more details to figure out what the underlying problem is? Does anyone have any pointers for me?

Cheers,
Finn

**Comment by webvictim** [comment, marked as answer]
Often this implies a request to the IP of the server rather than a hostname (like a TCP health check for example)

`autocert` (the Go package Teleport uses for its built-in ACME implementation) is pretty dense and just outputs that error whenever it sees a request without a server name.

One workaround to hide that specific message is to get certs from `certbot` and provide those to Teleport using the `https_keypairs` config section of `/etc/teleport.yaml` rather than using the built-in ACME.

Does your Teleport setup actually work?

**Reply by ErrorProne** [reply]
Thanks for the anwser.
I'm still in the middle of the setup guide, so no working teleport setup. I pretty much fail on the first time teleport tries to setup the certificates.

Based on what you've said I'll recheck the haproxy configuration and look for the SNI settings, maybe something is wrong there. (Edit: Yup, haproxy is currently offloading and re-initiating a new TLS request which breaks it)

If I fall back to certbot and let a cron refresh the certs, should the cert, set in the https_keypairs section, contain both the wildcard and the "normal" URI? So for example *.teleport.mydomain.com and teleport.mydomain.com.

**Reply by webvictim** [reply]
If you're looking to use application access via Teleport (which uses subdomains) then adding `*.teleport.example.com` as a SAN on the certbot request makes sense. If not, don't worry about it and just get a cert for `teleport.example.com`

To make HAProxy work you'll want to deploy it in TCP mode and let Teleport terminate its own TLS.

**Reply by ErrorProne** [reply]
Thanks for the explanation. I'll try to do some trickery to allow offloaded and tcp mode in parallel on the same port and if everything falls apart the certbot variant will be the fallback.

Anyway, thanks for the insights, 

> autocert (the Go package Teleport uses for its built-in ACME implementation) is pretty dense and just outputs that error whenever it sees a request without a server name.

was the key to understand what is going on. Thank you very much.
